proposals/shared-everything-threads/Overview.md

```diff
@@ -234,33 +370,291 @@ fingerprinting. Note that:
 2. this kind of fingerprinting is likely possible without the instruction via some careful timing of
    shared access from concurrently-executing threads
 
-This issue can be discussed further here: [Should we include
-`thread.hw_concurrency`?][hw-concurrency-discussion].
+Though this functionality was considered non-essential in previous
+[discussions][hw-concurrency-discussion], those were in the context of browsers, with the
+[`hardwareConcurrency`] property available. We expect standalone engines to need some equivalent for
+general parallel programming.
 
 [`hardwareConcurrency`]: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/hardwareConcurrency
 [widely available]: https://caniuse.com/?search=hardwareConcurrency
 [hw-concurrency-discussion]: https://github.com/abrown/thread-spawn/discussions/6
 
+### Managed Waiter Queues
+
+Linear memory Wasm has `memory.atomic.wait32`, `memory.atomic.wait64`, and `memory.atomic.notify`
+that provide a [futex-like interface][wait-notify] to runtime-managed waiter queues. The `wait`
+instructions allow threads to atomically check a value at a particular memory address and begin
+waiting if the value is what they expect. The `notify` instruction wakes some number of waiters
+associated with a particular memory address. WasmGC programs using threads will need a similar
+primitive that does not depend on linear memory. We propose adding a new abstract heap type
+`waitqueue` to serve as this new primitive.
+
+Unlike other abstract heap types, which are only ever subtypes of other abstract heap types (or are
+bottom types), `waitqueue` is a final subtype of `(rec (type (sub (struct shared (field i32))))).0`,
+meaning it is a struct with one visible `i32` field that can be accessed with all the standard
+struct accessors as well as the new atomic struct accessors. This field is the futex control field
```

Comment:
Another option might be to make `waitqueue` an attribute of an i32 field, rather than an entire struct with a singleton field. In theory it seems more cache-friendly to embed a waitqueue into a GC struct like that, rather than "box" it, but I'm guessing that was considered and has some downsides?

Reply:
So waitqueues are just futexes with a layer of indirection removed. Ordinarily, futexes are keyed by an address for its futex word, because that's the level of abstraction the OS is working at. Replicating the address-as-key API means not just the DX annoyance but also additional bookkeeping in implementations. Since any shared linear memory address can be a futex key, implementations need a threadsafe hashtable to manage the waitqueues. You can remove that level of indirection by providing waitqueues directly, since we now have the luxury of struct references and not only addresses.

With that mental model, boxing it is natural. The goal is to not require a threadsafe hash table that maps pairs of (holder struct, i32 field) keys to waitqueues. Making it an attribute on i32 fields seems like it could work, as long as implementations can represent the waitqueues to be in the holder structs themselves.